### PR TITLE
Add distinct options for general and dotprod attention

### PR DESCRIPTION
**Pull request by bpopeters**
The option for dotprod attention in GlobalAttention in fact matched Luong et al.'s specification of general attention. I fixed the naming of these options and added an attention option that corresponds to the dot product method of scoring in the Luong paper. 

I refactored the attention-scoring a little bit so now the batch x source length attention score tensor can be created by a helper score function. If the GlobalAttention class were only concerned with the mechanisms from Luong et al.'s paper, it would be possible to handle all of their attention types simply by subclassing the thing and defining a new score function and layers as necessary. However, the process for MLP has a few different steps later on in forward().

### CODE REVIEW

**Comment by jsenellart** [comment]
we should try to normalize option names and values with lua version. In lua version, we have `-attention (local|global|none)`, and `-attention_type general|dot|dot_scaled|concat` - where I think that `concat` is corresponding to `mlp` (Badhanau), and `dotprod` is `dot`. Since the python version is still being defined, I propose to align to the OpenNMT options.

**Comment by bpopeters** [comment]
I agree that trying to keep things consistent between lua and python would be useful. In the particular case of global attention, I'm not sure whether lua's `concat` is the same as the python `mlp`. Per the OpenNMT docs, their `concat` is the version from Luong's paper, which looks easier and more straightforward to do alongside the dot/general options than MLP. All that one needs to do is overwrite the `score` function.

**Comment by bpopeters** [comment]
Sorry about that, I fixed my whitespace problem now.

File: onmt/modules/GlobalAttention.py [hunk 1]
```diff
@@ -39,18 +39,17 @@ def __init__(self, dim, coverage=False, attn_type="dotprod"):
...
             self.linear_in = nn.Linear(dim, dim, bias=False)
-            self.linear_out = nn.Linear(dim*2, dim, bias=False)
-        elif self.attn_type == "mlp":
+        elif self.attn_type == 'mlp':
             self.linear_context = BottleLinear(dim, dim, bias=False)
             self.linear_query = nn.Linear(dim, dim, bias=True)
-            self.mlp_tanh = nn.Tanh()
             self.v = BottleLinear(dim, 1, bias=False)
-            self.linear_out = nn.Linear(dim*2, dim, bias=True)
+        out_bias = self.attn_type == 'mlp'  # mlp wants it with bias
```
Comment (srush) [inline onmt/modules/GlobalAttention.py]: Please put the comment on its own line.

File: onmt/modules/GlobalAttention.py [hunk 2]
```diff
@@ -39,18 +39,17 @@ def __init__(self, dim, coverage=False, attn_type="dotprod"):
...
-        assert (self.attn_type in ["dotprod", "mlp"]), (
+        assert (self.attn_type in ["dotprod", "general", "mlp"]), (
                 "Please select a valid attention type.")
 
-        if self.attn_type == "dotprod":
+        if self.attn_type == "general":
             self.linear_in = nn.Linear(dim, dim, bias=False)
-            self.linear_out = nn.Linear(dim*2, dim, bias=False)
-        elif self.attn_type == "mlp":
+        elif self.attn_type == 'mlp':
```
Comment (srush) [inline onmt/modules/GlobalAttention.py]: let's be consistent with ' versus "

File: onmt/modules/GlobalAttention.py [hunk 3]
```diff
@@ -39,18 +39,17 @@ def __init__(self, dim, coverage=False, attn_type="dotprod"):
 
         self.dim = dim
         self.attn_type = attn_type
-        assert (self.attn_type in ["dotprod", "mlp"]), (
+        assert (self.attn_type in ["dotprod", "general", "mlp"]), (
                 "Please select a valid attention type.")
 
-        if self.attn_type == "dotprod":
+        if self.attn_type == "general":
```
Comment (srush) [inline onmt/modules/GlobalAttention.py]: I think these names should be more clear. Maybe there can be a comment with the math for each of them? Alternatively we can sync with OpenNMT Lua

File: onmt/modules/GlobalAttention.py [hunk 4]
```diff
@@ -62,11 +61,36 @@ def __init__(self, dim, coverage=False, attn_type="dotprod"):
     def applyMask(self, mask):
         self.mask = mask
 
+    def score(self, h_t, h_s):
+        """
+        h_t (FloatTensor): batch x dim
```
Comment (srush) [inline onmt/modules/GlobalAttention.py]: Instead of a comment, can you use aeq() here to enforce this check.

**Comment by srush** [comment]
This looks great. A couple changes requested, and then I am happy to merge.

**Comment by bpopeters** [comment]
I adapted the Luong attention comments from GlobalAttention.lua and changed "dotprod" to "dot".

**Comment by bpopeters** [comment]
I forgot that changing "dotprod" to "dot" would create problems for the tests as well. I'm not sure that's something I can fix...

Also, "dot" might not be the best choice for default attention, since lua uses general (and historically, the thing that was called dotprod here was really general).
